UseCase Private Vault for Component

ferenc-hechler opened the pull request:
As anounced in issue #78 this pull request adds the usecase "Private Vault for Component".

gjo-satec commented:
I have some thoughts to share, but I must admit that I haven't yet found a one-size-fits-all solution to this problem.
Although I share the concerns about the kubernetes secrets, I also doubt about ad-hoc solutions
Using a custom sidecar for this task adds a lot of burdens:
1.- The sidecar itself. Having such an element implies that someone has to develop and maintain
2.- Although some vaults are compatible with OIDC, others, like AWS Secrets, are not, although tools like Iam for OIDC help to fill the gap  
3.- All the deployments have to know about that element . It implies to change code of the components itself

I aim more for a standard solution for these kinds of problems unless our requirements are so stringent that we have to use a custom solution.
If we want to get rid of the secrets, we can use [Secrets Store CSI Driver]( https://secrets-store-csi-driver.sigs.k8s.io/introduction.htm)l to mount the secrets in a tmpfs directly from a Vault Providers. It brings  its own limitations,but that it's another history
We also can use tools like [External Secrets](https://external-secrets.io/v0.8.2/) , kubernetes secrets and let the Canvas Operator create some stringent RBAC policies for the Service Account associate with an ODA deployment .Something like this example

````apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  namespace: my-namespace
  name: secret-reader
rules:
- apiGroups: [""]
  resources: ["secrets"]
  resourceNames: ["my-secret"] 
  verbs: ["get", "watch", "list"]
````
````apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: read-secret-binding
  namespace: my-namespace
subjects:
- kind: ServiceAccount
  name: my-service-account
  namespace: my-namespace
roleRef:
  kind: Role
  name: secret-reader
  apiGroup: rbac.authorization.k8s.io`
````
````
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-deployment
  namespace: my-namespace
spec:
  template:
    spec:
      serviceAccountName: my-service-account
      containers:
      - name: my-container
        image: my-image
 ''''
Adding a correct etcd configuration to add encryption at rest would be necessary as well

ferenc-hechler commented:
@gjo-satec, thanks for the feedback. As it is a proposal, I am happy to discuss my thoughts about it.   :-)

to 1) "The sidecar itself. Having such an element implies that someone has to develop and maintain"

This holds true for all decoupling components, like the operators. They are part of the Canvas Implementation and not part of the specification. 

The SideCar is thought as a lean adapter which translates the simple request into the specific Cavas-Vault which is used in the corresponding canvas implementation. So, the component can be developed relying on a simplified API and work together with any kind of Vault, that supports JWT authentication. Handling the JWT is not part of the Adapater. The adapter only knows, where to find the JWT in its filesystem and enrichs the request with this information.

to 2) "Although some vaults are compatible with OIDC, others, like AWS Secrets, are not, although tools like Iam for OIDC help to fill the gap"

As said above, the handling of JWT is up to the Canvas-Vault, the configuration of the Canvas-Vault is up to the Component-Operator. The SideCar is only translating the requests into the specific Canvas-Vault implmentation (Hashicorp Vault, CyberArk Conjur, IAM for OIDC, ...)

Of Course, the decision, which Canvas-Vault to use is restricted to implementations, which support authentication with a JWT.

to 3) "All the deployments have to know about that element . It implies to change code of the components itself"

That was not the idea behind. It is an option, which can be used for solving one hard issue which a component might faces. How can I store secrets securely. At the end It is sufficient to have one strong secret, which the component can rely on to have it exclusively, then any other kind of Password-Infrastructure can be built upon this.

Let my just explain my thoughts about this.

Let´s assume a Component creates an in-memory private key on startup which is used to encrypt other sensitive information. It can publish the public key and everything works fine. No one (even with full access to the cluster) can steel the private key. But the in-memory private-key gets lost, as soon as the POD restarts, so there is a need to persist this key.

And now we get to the inifinite loop of how to persist a secret in a secure manner. To use a private key to encrypt this key is a recursion. So, the next Idea is to use a standard software which was built having security in mind. No application specific security by obscurity. Great Idea. Maybe the component ships an implementation of, e.g. Hashicorp-Vault, and deploys it together with the component. But Suprise-Suprise, HashicorpVaul has the same issue to create an intial Secret. So, it has to be unsealed with a given number of secrets which have to be stored somewhere outside the cluster.

Okay, them let´s outsource this issue to the canvas implmentation. It is up to the canvas to setup and maintain a secure vault, we trust the Canvas-Vault to be secured. The problem to solve now is: "How to login to the Vault?"

The Component-Operator has the permissions to configure the Vault, it could create a derdicated UserPass Login for the component, but how to tell the component implementation the password? As an environment variable or Kubernetes Secret breaks our whole security chain.

And FINALLY, there is a solution to break the recursion. The Kubernetes cluster itself can act as an authority which proves the identity of a POD and the ServiceAccount. So, Component-Operator creates a JWT based login, which can be used only by the component instance, it was created for. 

Finally I had the decoupling in mind, because the canvas can use arbitrary Vault implementations and I did not want to hard code one implementation. Therefore the sidecar was added. In an older version I described another workflow, where the component implementation knows about the JWT in the file-system and uses it to communicate with the Canvas-Vault. If you like, i can add this as an alternative Workflow again.

![grafik](https://github.com/tmforum-oda/oda-canvas/assets/11094156/8c76e4b3-bb35-4be8-8980-0383fce26533)

ferenc-hechler commented:
As I understood Dave Milham has also planned to discuss this proposal in the "ODA Security & Privacy Meeting" Wednesday 13:00 CET. 
But if this fits not, I can also join your meeting on Thursday.

pecanpete commented:
Ues as Ferenc mention ed we have asked him to present on next week's OD Sec and Privacycall Wed starting 12:00 midday BST and will run for two hours . It notion of secret management was raised a few weeks ago by Brad Peters NBN and we want to include it in the ODA supporting Function/Canvas specs and obviously want it aligned with ODA CA thinking too.

ferenc-hechler commented:
> Hi @ferenc-hechler - do you think it is worth having a call to discuss this proposal? We could either invite you to our regular working group on Thursday at 12 noon CET, or have a specific call for this. I believe @brian-burton as well as @gjo-satec would be interested in discussing the options.

@LesterThomas:
Discussed the workflow today in the "ODA Security & Privacy Meeting". Main feedback was, that the whole process relies in our trust to the Kubernetes implementation. 
I am happy to discuss with you tomorrow.  :-)
Can you please invite me.

brian-burton commented:
> PING: I have no invitation for the meeting today. Can someone tell me, where to find it.

Hi @ferenc-hechler, I've asked for the topic to be scheduled and an invite sent out to you.

@pecanpete please can you check your inbox and help facilitate that?

ferenc-hechler commented:
I would like to redesign the Workflow to a more generic one, which is not directly JWT based.
The reason is, that we are using Conjur with Certificate based authentication in Rancher.
The abstract Workflow steps are the same, but have to be formulated more generic.

I would also add the current JWT diagram as an example implementation. 
The Implementation would then be using the open source HashiCorp Vault.

pecanpete commented:
Feri  Actually I like your proposal as this abstraction is exactly what we need in the ODA Cnavas and ODA Component Supporting Function work in SF001 
https://projects.tmforum.org/wiki/pages/viewpage.action?pageId=221382600&src=contextnavpagetreemode
This work is the abstraction of the ODA CA work and is in TMF RAND mode rather than Apache 2. Current consent scope  is IAM but I am wondering that as this work in not due until end Sprint 4 whether introducing Secrets Vault would be good stretch target . Note the current document needs some refactoring inn to three parts ( to dierectly support procurement) and a discussion for  the ODA Canvas team calls

ferenc-hechler commented:
Added the description for the generic workflow, which is not depending on JWT:
https://github.com/ferenc-hechler/oda-canvas/blob/master/usecase-library/UC0XX-Component-Vault.md

ferenc-hechler commented:
PING: Before continuing to work on the Operator, I would need some feedback if I am going into the right direction...

ferenc-hechler commented:
We had an internal ODA-Workshop and discussed the current state of the PrivateVault.
Here is the feedback:

## Allow other kind of Secret access

The current description only describes the access via an local API, which needs a sidecar `type: sidcar`.
For some standard components it might not be possible to access the secrets in this way.
Therefore other access methods should be possible:

* `type: sidecar` the secrets are accessible via a local API as described above
* `type: file` the secrets are injected into a memory filesystem by an init-container
* `type: file-update` the secrets are injected and updated into a memory filesystem by a sidecar
* `type: env` the secrets are injected and updated by the PrivateVault-Operator as Environment variables in the POD (POD manifest contains plain values)
* `type: env-entrypoint` the POD gets an entrypoint injected, which sets the environment variables (maybe technically difficult)
* `type: k8s-secret` the secrets are made available in the namespace as kubernetes secrets at POD starting time 
* `type: k8s-secret-update` the secrets are made available in the namespace and updated regularly

## Allow to define PODS with readonly access to PrivateVault

In the current concept, only one role with Full-Access to the PrivateVault Keystore is created.
It might be good to define more fine granular permissions, like PODs which are allowed to read, but not to write secrets.

Thinking more about this, it could be solved by a generic definition of roles with own settings

```
security:
  ...
  privateVault:
    - name: admin
      permissions: ["create", "read", "update", "delete", "list", "patch"]
      selector: 
        podnames:
        - db-mgmt
    - name: read
      permissions: ["read"]
      selector: 
        podnames:
        - db-server
        - db-client
    - name: setup
      permissions: ["create"]
      selector: 
        podnames:
        - db-init
````

## Allow initial creation of Secrets

It should be possible to generate secrets with an initial value.
So, nowhere in the code the secret is visible, but on initial deploy time a value is generated and can be used on server and client side

## Key Rotation

Allow periodic key rotations. 


## Conjur Support

As DT is using CyberArk Conjur and not HashiCorp Vault, a reference Implementation for Conjur would also be appreciated.
